File: android/translate-kit/src/main/java/io/github/marcosholgado/translatekit/TranslateKit.kt
/*
 * Copyright 2026 Marcos Holgado
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.marcosholgado.translatekit

import android.content.Context
import android.util.Log

/**
 * Entry point to translate-kit: on-device neural machine translation + offline
 * language detection.
 *
 * Usage mirrors a small facade with a lazy [init] driven from the caller's IO
 * dispatcher. All native calls are **blocking**; never call them on the main
 * thread.
 *
 * The library never touches the network. Translation models are supplied by the
 * caller as on-disk paths ([loadModel]); the language detector is bundled and
 * self-contained (no model file to load).
 */
object TranslateKit {

    private const val LIBRARY_NAME = "translate-kit"
    private const val TAG = "TranslateKit"

    @Volatile
    private var nativeContextPtr: Long = 0L

    @Volatile
    private var nativeLibraryUnavailable: Boolean = false

    /**
     * Loads the native library and initializes the engine. Idempotent and
     * crash-safe. Blocking — call from an IO dispatcher.
     *
     * translate-kit ships native libraries for `arm64-v8a` and `x86_64` only. A
     * host app may also ship 32-bit ABIs (`armeabi-v7a`, `x86`), so this can run
     * on a device with no matching `.so`. In that case it degrades gracefully —
     * it does **not** crash the app (an `UnsatisfiedLinkError` is an [Error],
     * which a caller's `Exception` handler would not catch); instead
     * [isInitialized] stays `false` and translation is unavailable on that
     * device. Gate the translation feature on [isInitialized].
     */
    @Synchronized
    fun init(context: Context) {
        if (nativeContextPtr != 0L || nativeLibraryUnavailable) return
        try {
            System.loadLibrary(LIBRARY_NAME)
        } catch (e: UnsatisfiedLinkError) {
            // No native library for this device's ABI. Degrade gracefully rather
            // than crashing the host app; translation stays unavailable here.
            nativeLibraryUnavailable = true
            Log.w(TAG, "translate-kit native library unavailable for this device's ABI; " +
                "on-device translation is disabled here", e)
            return
        }
        val ptr = nativeInit()
        check(ptr != 0L) { "translate-kit init failed: ${nativeLastError()}" }
        nativeContextPtr = ptr
    }

    /**
     * `true` once [init] has loaded the native library and initialized the
     * engine. `false` before [init], or on a device whose ABI translate-kit
     * does not ship a native library for (see [init]).
     */
    fun isInitialized(): Boolean = nativeContextPtr != 0L

    /** Offline language detection using the bundled model. Blocking. */
    fun detectLanguage(text: String): LanguageResult {
        val ptr = nativeContextPtr
        check(ptr != 0L) { "TranslateKit.init(context) must be called before detectLanguage" }
        return nativeDetectLanguage(ptr, text)
    }

    /**
     * Load a translation model from on-disk files. Returns a handle used for
     * subsequent [TranslationModel.translate] calls. Throws on bad/missing
     * files. Blocking.
     */
    fun loadModel(spec: ModelSpec): TranslationModel {
        val ptr = nativeContextPtr
        check(ptr != 0L) { "TranslateKit.init(context) must be called before loadModel" }
        val modelPtr = nativeLoadModel(
            ptr,
            spec.sourceLang,
            spec.targetLang,
            spec.modelPath,
            spec.vocabPaths.toTypedArray(),
            spec.shortlistPath,
            spec.configYaml,
            spec.numWorkers,
        )
        if (modelPtr == 0L) {
            throw IllegalArgumentException("translate-kit loadModel failed: ${nativeLastError()}")
        }
        return TranslationModel(modelPtr)
    }

    /** Releases a loaded model. Equivalent to [TranslationModel.close]. */
    fun unloadModel(model: TranslationModel) {
        model.close()
    }

    private external fun nativeInit(): Long

    private external fun nativeDetectLanguage(nativeContextPtr: Long, text: String): LanguageResult

    private external fun nativeLoadModel(
        nativeContextPtr: Long,
        sourceLang: String,
        targetLang: String,
        modelPath: String,
        vocabPaths: Array<String>,
        shortlistPath: String,
        configPath: String?,
        numWorkers: Int,
    ): Long

    private external fun nativeLastError(): String
}

File: apple/Sources/TranslateKit/TranslateKit.swift
/*
 * Copyright 2026 Marcos Holgado
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import CTranslateKit

#if canImport(Glibc)
import Glibc
#else
import Darwin
#endif

/// Entry point to translate-kit: on-device neural machine translation + offline
/// language detection. Mirrors the Android `TranslateKit` facade and SPEC-1 §5,
/// over the same platform-neutral C ABI.
///
/// An instance owns one engine context (`tk_context`); its `deinit` tears it
/// down. The bundled language detector (CLD2) is self-contained, so no model
/// file is needed for ``detectLanguage(_:)``. Translation models are supplied by
/// the caller as on-disk paths via ``loadModel(_:)``. The library never touches
/// the network.
///
/// - Important: All calls are **blocking** — call them off the main thread. A
///   `TranslateKit` instance and the ``TranslationModel``s loaded from it are
///   **not** thread-safe: use one from a single background thread at a time
///   (serialize access, e.g. via a dedicated queue or actor).
public final class TranslateKit {

    private let context: OpaquePointer

    /// The library version string, e.g. `"0.1.0"`.
    public static var version: String { String(cString: tk_version()) }

    /// Initializes the engine. Throws if native initialization fails. Blocking.
    public init() throws {
        var context: OpaquePointer?
        try tkCheck(tk_init(&context))
        guard let context else {
            throw TranslateKitError.engine(String(cString: tk_last_error()))
        }
        self.context = context
    }

    deinit {
        tk_shutdown(context)
    }

    /// Detects the language of `text` using the bundled detector. Blocking.
    public func detectLanguage(_ text: String) throws -> LanguageResult {
        var result = tk_language_result()
        let status = text.withCString { tk_detect_language(context, $0, &result) }
        try tkCheck(status)
        let language = withUnsafeBytes(of: result.language) { raw in
            String(cString: raw.bindMemory(to: CChar.self).baseAddress!)
        }
        return LanguageResult(language: language, confidence: result.confidence)
    }

    /// Loads a translation model from on-disk files. Throws on bad/missing
    /// files. Blocking.
    public func loadModel(_ spec: ModelSpec) throws -> TranslationModel {
        // Own every C string for the duration of the call, then hand the C ABI
        // an array of stable pointers (mirrors the Android JNI shim).
        var allocations: [UnsafeMutablePointer<CChar>] = []
        func dup(_ string: String) -> UnsafePointer<CChar> {
            let copy = strdup(string)!
            allocations.append(copy)
            return UnsafePointer(copy)
        }
        defer { for pointer in allocations { free(pointer) } }

        var cSpec = tk_model_spec()
        cSpec.source_lang = dup(spec.sourceLang)
        cSpec.target_lang = dup(spec.targetLang)
        cSpec.model_path = dup(spec.modelPath)
        cSpec.shortlist_path = dup(spec.shortlistPath)
        cSpec.config_path = spec.configYaml.map(dup)
        cSpec.num_workers = Int32(max(1, spec.numWorkers))

        let vocabPointers: [UnsafePointer<CChar>?] = spec.vocabPaths.map(dup)
        var model: OpaquePointer?
        let status: tk_status = vocabPointers.withUnsafeBufferPointer { buffer in
            cSpec.vocab_paths = buffer.baseAddress
            cSpec.vocab_count = buffer.count
            return withUnsafePointer(to: &cSpec) { tk_load_model(context, $0, &model) }
        }
        try tkCheck(status)
        guard let model else {
            throw TranslateKitError.modelLoad(String(cString: tk_last_error()))
        }
        return TranslationModel(handle: model)
    }

    /// Releases a loaded model. Equivalent to ``TranslationModel/close()``.
    public func unloadModel(_ model: TranslationModel) {
        model.close()
    }
}
